creusot/tests/should_succeed/iterators/15_enumerate.rs

```diff
@@ -0,0 +1,83 @@
+#![feature(slice_take)]
+extern crate creusot_contracts;
+
+use creusot_contracts::*;
+
+mod common;
+use common::Iterator;
+
+pub struct Enumerate<I> {
+    iter: I,
+    count: usize,
+}
+
+impl<I: Iterator> Enumerate<I> {
+    #[predicate]
+    fn invariant_aux(self) -> bool {
+        pearlite! {
+            self.iter.invariant()
+            && forall<s: Seq<I::Item>, i: I> i.invariant() ==> self.iter.produces(s, i) ==> @self.count + s.len() < @std::usize::MAX
+        }
+    }
+}
+
+impl<I> Iterator for Enumerate<I>
+where
+    I: Iterator,
+{
+    type Item = (usize, I::Item);
+
+    #[predicate]
+    fn completed(&mut self) -> bool {
+        pearlite! { self.iter.completed() }
+    }
+
+    #[predicate]
+    fn produces(self, visited: Seq<Self::Item>, o: Self) -> bool {
+        pearlite! {
+            visited.len() == @o.count - @self.count
+            && exists<s: Seq<I::Item>> self.iter.produces(s, o.iter)
+                && visited.len() == s.len()
+                && forall<i: Int> 0 <= i && i < s.len() ==> @visited[i].0 == @self.count + i && visited[i].1 == s[i]
+        }
+    }
+
+    #[predicate]
+    fn invariant(self) -> bool {
+        pearlite! {
+            self.invariant_aux()
+            && forall<reset : &mut Enumerate<I>> reset.completed() ==> (^reset).invariant_aux()
```

Comment:
I'm not sure that this is provable. `invariant_aux` requires `(^reset).invariant()` but I don't think you can prove that here.

Reply:
I think you'll really need to actually build an `enumerate` iterator to see whether this invariant is actually satisfiable.